clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2108,9 +2146,12 @@ void handleFailure(Throwable throwable) {
     public DescribeTopicsResult describeTopics(final TopicCollection topics, DescribeTopicsOptions options) {
         if (topics instanceof TopicIdCollection)
             return DescribeTopicsResult.ofTopicIds(handleDescribeTopicsByIds(((TopicIdCollection) topics).topicIds(), options));
-        else if (topics instanceof TopicNameCollection)
+        else if (topics instanceof TopicNameCollection) {
+            if (options.useDescribeTopicsApi()) {
+                return DescribeTopicsResult.ofTopicNameIterator(new DescribeTopicPartitionsIterator(((TopicNameCollection) topics).topicNames(), options));
```

Comment:
DescribeTopicsResult is a Map based results class, which does not align the streaming/paging approach we are trying to achieve with this new RPC. I don't think we should try to reuse DescribeTopicsResult for the new API.

Something like
```
KafkaAdminClient#describeTopics(Consumer<TopicDescription>);
```

would be more natural for a streaming interface. With an interface like this, we just need to be concerned with efficiently paging through the results. The application can gather/filter/post-process the results as needed. 

Since this is a new public API, and really the first time doing streaming in KafkaAdminClient, I do think we should bring this up in the mailing list to collect feedback. E.g., do we want to return an Iterator/Stream, or accept a function like shown above.

Reply:
@mumrah If we need to spend more time on public discussion, maybe we can first have a simple version like paging on a topic level. Also, limit the number of topics loading in memory like maybe 50 to accomplish the short-term goal of avoiding OOM.
cc @artemlivshits

Reply:
It's been my experience that it's "dangerous" 😨 to run arbitrary user code from within the context of the client code. User code can (and will) do unpredictable things with state, errors, threads, etc. The surrounding code inside the client has to be very careful to make sure it handles many different cases.

That said, I like the ergonomics of the `Consumer`-based approach over the iterator.